Jenkinsfile

```diff
@@ -1,125 +1,142 @@
+        stage('Run unit tests') {
             steps {
-                // Direct the HOME to WORKSPACE for pip to get the
-                // installed library.
-                // 'PATH' can only be updated in a single shell block.
-                // We can not update PATH in 'environment' block.
-                // Pytest needs to export the junit report.
                 withEnv(["HOME=${env.WORKSPACE}"]) {
                     sh """
-                        source /home/saluser/.setup_dev.sh || echo loading env failed. Continuing...
-                        setup -k -r .
-                        pytest --cov-report html --cov=${env.MODULE_NAME} --junitxml=${env.XML_REPORT}
+                        source /home/saluser/.setup_dev.sh || echo "Loading env failed; continuing..."
+                        setup -r .
+                        pytest --cov-report html --cov=${env.MODULE_NAME} --junitxml=${env.XML_REPORT_PATH}
                     """
                 }
             }
         }
-        stage('Build and Upload Documentation') {
+        stage('Build documentation') {
             steps {
                 withEnv(["HOME=${env.WORKSPACE}"]) {
                     sh """
-                        source /home/saluser/.setup_dev.sh || echo loading env failed. Continuing...
-                        pip install .
+                        source /home/saluser/.setup_dev.sh || echo "Loading env failed; continuing..."
+                        setup -r .
                         package-docs build
-                        ltd upload --product ts-watcher --git-ref ${GIT_BRANCH} --dir doc/_build/html
                     """
                 }
             }
         }
+        stage('Try to upload documentation') {
+            steps {
+                withEnv(["HOME=${env.WORKSPACE}"]) {
+                    catchError(buildResult: 'UNSTABLE', stageResult: 'UNSTABLE') {
+                        sh '''
+                            source /home/saluser/.setup_dev.sh || echo "Loading env failed; continuing..."
+                            setup -r .
+                            ltd -u ${LSST_IO_CREDS_USR} -p ${LSST_IO_CREDS_PSW} upload \
+                                --product ${DOC_PRODUCT_NAME} --git-ref ${GIT_BRANCH} --dir doc/_build/html
+                        '''
+                    }
+                }
+            }
+        }
     }
-
     post {
         always {
-            // Change the ownership of workspace to Jenkins for the clean up
-            // This is a "work around" method
+            // Change ownership of the workspace to Jenkins for clean up.
             withEnv(["HOME=${env.WORKSPACE}"]) {
                 sh 'chown -R 1003:1003 ${HOME}/'
             }
 
-            // The path of xml needed by JUnit is relative to
-            // the workspace.
+            // The path of xml needed by JUnit is relative to the workspace.
             junit 'jenkinsReport/*.xml'
```

Comment:
Could be deduced from XML_REPORT_PATH?  Something like junit File($XML_REPORT_PATH).getParentFile().getName() + '/*.xml' ?

Reply:
This is from our official Jenkinsfile template in tssw_developer_guide. If you have any suggested improvements then it'd be great for T&S if you made them to that package.

I confess I'm not sure why you think it would be better to change this line. It looks nice and simple to me.